### DISCUSSION TITLE
Receive data from javascript websocket

### DISCUSSION THREAD

**Discussion by Omiros99** [micropython/micropython#9344] [Libraries & Drivers]
Hello everybody,

I am trying to receive some data from client side javascript (no node.js) but i can't make it work with the websocket libraries that i found 
(
    https://git.k-space.ee/lauri/micropython-skeleton/src/branch/master/uwebsockets.py ,
    https://github.com/jczic/MicroWebSrv2
)

Here is the javascript code:

```js
const ws = new WebSocket("wss://192.168.4.1:82");
ws.binaryType = 'arraybuffer';
ws.send("Hello");
ws.close(1000, "Done");
```

And the micropython code:

```py
def receiveData():
        #CONNECTED = True
        global CONNECTED
        
        sServer = socket.socket()
        sServer.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sServer.bind(('', 82))
        sServer.listen(16)
        while CONNECTED:
            conn, addr = sServer.accept()
            response = conn.recv(1024)
            if(response != b' '):
                print(response)
                CONNECTED = False
        sServer.close()
        return response
```

Of course i don't want to send a "Hello" message, this is just for testing purposes.
Even though i don't use any library for micropython websockets i receive some kind of data:
```py
b'\x16\x03\x01\x02\x00\x01\x00\x01\xfc\x03\x03\x8bD\\\xe3\xcf@`\x7f\r\xb0D\xfb\x01\x03y\x16\x99\xd2\xd0\xd0(\x034\xa9\xf7f\x90F\x14\x06G# )\x999R\xacG\x9c\n\xfc,\xe3\xda\x04\xc0\x84\x9b\x1bTF4P\xc9\xc4\xa5\xd8 \xe1b\x017\x03}\x00 JJ\x13\x01\x13\x02\x13\x03\xc0+\xc0/\xc0,\xc00\xcc\xa9\xcc\xa8\xc0\x13\xc0\x14\x00\x9c\x00\x9d\x00/\x005\x01\x00\x01\x93\xda\xda\x00\x00\x00\x17\x00\x00\xff\x01\x00\x01\x00\x00\n\x00\n\x00\x08**\x00\x1d\x00\x17\x00\x18\x00\x0b\x00\x02\x01\x00\x00#\x00\x00\x00\x10\x00\x0b\x00\t\x08http/1.1\x00\x05\x00\x05\x01\x00\x00\x00\x00\x00\r\x00\x12\x00\x10\x04\x03\x08\x04\x04\x01\x05\x03\x08\x05\x05\x01\x08\x06\x06\x01\x00\x12\x00\x00\x003\x00+\x00)**\x00\x01\x00\x00\x1d\x00 _\x02_*\x8f$\x86\xda\x91N\x9ev8P\xabIvfj\x15#\xd2M^\xc5\xc0#\xe1\xedM\x1f\x1c\x00-\x00\x02\x01\x01\x00+\x00\x07\x06\xfa\xfa\x03\x04\x03\x03\x00\x1b\x00\x03\x02\x00\x02Di\x00\x05\x00\x03\x02h2zz\x00\x01\x00\x00\x15\x00\xe3\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
```

Does anybody has any idea how i will be able to make this work? I feel that i am so close to finish this task, any kind of help could be really really helpful.

Thank you in advance.

**Comment by jimmo** [comment]
FWIW, the webrepl uses websockets with thr built-in websocket module.
https://github.com/micropython/micropython-lib/blob/master/micropython/net/webrepl/webrepl.py might give you some hints.

**Reply by Omiros99** [reply]
I think that the build in module for websockets is called uwebsocket and as i remember it wasn't very useful in my case, but anyway i will give it a try. Thank you very much for your answer.

**Comment by ThinkTransit** [comment]
I would recommend microdot as they recently introduced websockets, including an async version.

Check out the simple example for a websocket echo server.

https://github.com/miguelgrinberg/microdot/blob/main/examples/websocket/echo.py

Also you are likely to have issues with SSL so maybe try and get it working with non ssl "ws:" first.

**Comment by MrTinkerman** [comment]
I'm a bit of a beginner at this but I got TCP socket communication working. Not sure if it's gonna help you but I'm using:
`
conn, address = server_socket.accept()
            data = conn.recv(128).decode()
`
It's just a guess but maybe ".decode()" could do the trick?

Here's my receiving code:
```
def listen():
    server_socket.bind((ip_address, socket_port))
    server_socket.listen(1)

    print(f'Listening on {ip_address} port {socket_port}')

    while True:
        print('>>>Waiting for connection')
        data = ''
        try:
            conn, address = server_socket.accept()
            data = conn.recv(128).decode()
            conn.send(data)
            conn.close()
        except OSError as error:
            print(error)
            listen()
        finally:
            pass

        if (data):
            command = json.loads(data)
            print(command)
            target = command['TARGET']

            if target == 'LED':
                set_led(command['VAL'])
            if target == 'MOTOR':
                turn_motor(command['DIR'], command['TIME_MS'])```

**Reply by Omiros99** [reply]
First of all, thank you for answering, but, Web Socket Protocol is layered over a TCP connection so is something like a "level-up" version of a TCP protocol. If you want to learn more i would suggest you to read this https://www.rfc-editor.org/rfc/rfc6455 it was an eye opener for me. Cheers and never stop grinding!

**Reply by MrTinkerman** [reply]
Thanks for the info, I have some reading to do :)
Good luck on your project!
